Libplanet.Tests/Blockchain/BlockChainTest.cs

```diff
@@ -704,30 +704,67 @@ public void UnstageAfterAppendComplete()
+            IActionRenderer<DumbAction> renderer = new AnonymousActionRenderer<DumbAction>
+            {
+                BlockRenderer = (oldTip, newTip) =>
+                   blockLogs.Add((oldTip, newTip, idx++)),
+                ActionRenderer = (act, context, nextStates) =>
+                    actionLogs.Add((new ActionEvaluation(act, context, nextStates), idx++)),
+            };
+            renderer = new LoggedActionRenderer<DumbAction>(renderer, Log.Logger);
+            BlockChain<DumbAction> blockChain =
+                TestUtils.MakeBlockChain(policy, store, renderers: new[] { renderer });
+            var privateKey = new PrivateKey();
+
+            var action = new DumbAction(default, string.Empty);
+            var actions = new[] { action };
+            blockChain.MakeTransaction(privateKey, actions);
+            idx = 0;
+            blockLogs.Clear();
+            actionLogs.Clear();
+            Block<DumbAction> prevBlock = blockChain.Tip;
+            Block<DumbAction> block = await blockChain.MineBlock(_fx.Address1);
+
+            Assert.Equal(2, blockChain.Count);
+            Assert.Single(blockLogs);
+            Assert.Single(actionLogs);
+            Assert.Equal((prevBlock, block, 0), blockLogs[0]);
+            Assert.Equal(1, actionLogs[0].Index);
+            Assert.Equal(action, actionLogs[0].Evaluation.Action);
+        }
+
         [Fact]
         public async Task RenderActionsAfterAppendComplete()
         {
-             var policy = new NullPolicy<DumbAction>();
-             var store = new DefaultStore(null);
-             IActionRenderer<DumbAction> renderer = new AnonymousActionRenderer<DumbAction>
-             {
-                 ActionRenderer = (_, __, nextStates) =>
-                     throw new SomeException("thrown by renderer"),
-             };
-             renderer = new LoggedActionRenderer<DumbAction>(renderer, Log.Logger);
-             BlockChain<DumbAction> blockChain =
-                 TestUtils.MakeBlockChain(policy, store, renderers: new[] { renderer });
-             var privateKey = new PrivateKey();
-
-             var action = new DumbAction(default, string.Empty);
-             var actions = new[] { action };
-             blockChain.MakeTransaction(privateKey, actions);
-
-             SomeException e = await Assert.ThrowsAsync<SomeException>(
-                 async () => await blockChain.MineBlock(_fx.Address1)
-             );
-             Assert.Equal("thrown by renderer", e.Message);
-             Assert.Equal(2, blockChain.Count);
+            var policy = new NullPolicy<DumbAction>();
+            var store = new DefaultStore(null);
+            IActionRenderer<DumbAction> renderer = new AnonymousActionRenderer<DumbAction>
+            {
+                ActionRenderer = (_, __, nextStates) =>
+                    throw new SomeException("thrown by renderer"),
+            };
+            renderer = new LoggedActionRenderer<DumbAction>(renderer, Log.Logger);
```

Comment:
Isn't it overwritten?

Reply:
That's the way to use `LoggedActionRenderer<T>` as it uses [decorator pattern][1].  See also the [example code][2] in its docs.

[1]: https://en.wikipedia.org/wiki/Decorator_pattern
[2]: https://docs.libplanet.io/main/api/Libplanet.Blockchain.Renderers.LoggedActionRenderer-1.html#Libplanet_Blockchain_Renderers_LoggedActionRenderer_1_examples